src/pstl_offload.cpp

```diff
@@ -113,43 +126,223 @@ __original_msize(void* __user_ptr)
     return __orig_msize(__user_ptr);
 }
 
+static auto
+__get_original_realloc()
+{
+    using __realloc_func_type = void* (*)(void*, std::size_t);
+
+    static __realloc_func_type __orig_realloc = __realloc_func_type(dlsym(RTLD_NEXT, "realloc"));
+    return __orig_realloc;
+}
+#elif _WIN64
+
+static __free_func_type __original_free_ptr = free;
+
+#endif
+
+inline bool
+__is_ptr_page_aligned(void* __p)
+{
+    // using that __get_memory_page_size() returns only power of 2 values
+    return ((std::uintptr_t)__p & (__get_memory_page_size() - 1)) == 0;
+}
+
+struct __hash_aligned_ptr
+{
+    std::uintptr_t operator()(void* __p) const
+    {
+        // We know that addresses are at least page-aligned, so, expecting page at least
+        // 4K-aligned, drop 11 right bits that are zeros, and treat rest as a pointer,
+        // hoping that an underlying Standard Library support this well.
+        static constexpr unsigned __ptr_shift = 11;
+        // current page size is at least 4K
+        assert(__get_memory_page_size() >= (1 << __ptr_shift));
+        return std::hash<void*>()((void*)((std::uintptr_t)__p >> __ptr_shift));
+    }
+};
+
+template<class T>
+struct __orig_free_allocator
+{
+    typedef T value_type;
+
+    __orig_free_allocator() = default;
+
+    template<class U>
+    constexpr __orig_free_allocator(const __orig_free_allocator <U>&) noexcept {}
+
+    T* allocate(std::size_t __n)
+    {
+        if (T *ptr = static_cast<T*>(std::malloc(__n * sizeof(T))))
+        {
+            return ptr;
+        }
+        throw std::bad_alloc();
+    }
+
+    void deallocate(T* __ptr, std::size_t) noexcept
+    {
+#if __linux__
+        __original_free(__ptr);
+#elif _WIN64
+        __original_free_ptr(__ptr);
+#endif
+    }
+};
+
+template<class T, class U>
+bool operator==(const __orig_free_allocator <T>&, const __orig_free_allocator <U>&) { return true; }
+template<class T, class U>
+bool operator!=(const __orig_free_allocator <T>&, const __orig_free_allocator <U>&) { return false; }
+
+static __spin_mutex __large_aligned_ptrs_map_mtx;
+
+class __large_aligned_ptrs_map
+{
+public:
+    struct __ptr_desc
+    {
+        __sycl_device_shared_ptr _M_device;
+        std::size_t _M_requested_number_of_bytes;
+    };
+
+private:
+    // Find sycl::device and requested size by user pointer. Use __orig_free_allocator to not
+    // call global delete during delete processing (overloaded global delete includes
+    // __unregister_ptr call).
+    using __map_ptr_to_object_prop =
+        std::unordered_map<void*, __ptr_desc, __hash_aligned_ptr, std::equal_to<void*>,
+        __orig_free_allocator<std::pair<void* const, __ptr_desc>>>;
+    __map_ptr_to_object_prop* _M_map;
```

Comment:
Could you please try the approach with 
```cpp
union {
    __map_ptr_to_object_prop _M_map;
}
```
and then *placement*-`new` here and let me know if it works. Then we can decide what we prefer

Reply:
One (pretty marginal) point here is that we need a to use `__large_aligned_ptrs_map` methods before its ctor run. With pointer-based implementation, this flag is just pointer. For union-based, some additional flag is required.